design/KEPS/namespace-discovery/proposal.md

```diff
@@ -12,19 +12,116 @@ Provide a better way for namespace discovery within both server and operator.
+
+Controlplane A discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Controlplane B discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+Kiali discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+  - matchLabels:
+      team: api
+```
+
+For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selector's defined on that cluster.
+
+Cluster A discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Cluster B discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+Kiali discovery selector cluster A:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Kiali discovery selector cluster B:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+The same rules for multiple revisions in a multi-cluster deployment would still apply.
+
+Primary-remote deployments would behave the same as single cluster where the remote cluster is scoped to the primary controlplane's discovery selectors.
+
+TODO: external controlplane
+
+### Failure scenarios
+
+Since the discovery selectors act as a way to separate multiple Kiali deployments from one another, when Kiali relies on autodiscovering the Istio discovery selectors and it cannot auto discover them, the Kiali server will fail to start and/or panic rather than potentially show out of scope data. In a multi-primary deployment, if one primary cannot be reached then that cluster will be considered inaccessible.
+
+## Discovery Selector Kiali Config Option
+
+A new configuration option will be added that will mimic the Istio discovery selectors configuration option. Using the configuration option, users will be able to specify which namespaces Kiali should be scoped to. Having a config option within Kiali has the advantage of not relying on autodiscovery of the Istio controlplanes. In cases where Kiali temporarily cannot access the control plane configuration, Kiali does not need to fail since it will rely on its own discovery selectors to scope itself. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.
+
+Kiali would add a top level `discoverySelectors` config option where you can specify a `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to.
```

Comment:
I don't think it should be top level - seems appropriate under "deployment" - where "accessible_namespaces" are today.

Reply:
I wasn't sure the best place for it. `Deployment` seems to be config options related to how Kiali is deployed. The other options in there are stuff like `InstanceName`, `Namespace`, etc. all things related to the actual Kiali deployment. Whereas this option doesn't have anything to do with how Kiali itself is deployed. I think `accessible_namespaces` made sense there because the operator would change how Kiali was deployed based on that field but that is what `cluster_wide_access` is now for. I'm happy to change this though.

Reply:
Could go under "istio" since this is supposed to declaring the mesh namespaces, but I don't want to add one more thing there :)

`deployment` is still applicable in the sense that it will determine where the Roles will be deployed (just like accessible namespaces).

This just doesn't seem a large enough "thing" to warrant being its own top-level setting. And since it kinda replaces accessible_namespaces, seems like it will be less confusing to put it under the  same category.

I am not strongly of opinion either way though :)

Make sure its called "discovery_selectors" though (snake_case, not camelCase) to remain consistent with everything else. :)